src/gemdat/transitions.py

```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+                                    source=start,
+                                    target=end,
+                                    weight='weight')
+    path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
+
+    return [optimal_path, path_energy]
+
+
+def find_best_perc_path(
+        F: np.ndarray,
+        peaks: np.ndarray,
+        percolate_x=True,
+        percolate_y=False,
+        percolate_z=False) -> typing.Union[float, tuple, tuple, np.ndarray]:
+    """Calculate the best percolating path.
+
+    Parameters
+    ----------
+    F : np.ndarray
+        Energy grid that will be used to calculate the shortest path
+    peaks : np.ndarray
+        List of the peaks that correspond to high probability regions
+    percolate_x : bool
+        If True, consider paths that percolate along the x dimension
+    percolate_y : bool
+        If True, consider paths that percolate along the y dimension
+    percolate_z : bool
+        If True, consider paths that percolate along the z dimension
+
+    Returns
+    -------
+    total_energy_cost: float
+        Total energy cost of the best percolating path
+    best_starting_point: tuple
+        Coordinates of the starting site where the best percolating path was found
+    best_perc_path: list
+        List of coordinates of the path
+    best_perc_path_energy: list
+        Energy along the path
+    """
+    X_real, Y_real, Z_real = F.shape
+
+    # Find percolation using virtual images along the required dimensions
+    if not (percolate_x + percolate_y + percolate_z):
+        print('Warning: percolation is not defined')
+        return None, None, None
+    if percolate_x:
+        extended_F = np.empty((2 * F.shape[0], F.shape[1], F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:X_real, :, :] = F
+        extended_F[X_real:, :, :] = F
+        F = extended_F
+    if percolate_y:
+        extended_F = np.empty((F.shape[0], 2 * F.shape[1], F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:, :Y_real, :] = F
+        extended_F[:, Y_real:, :] = F
+        F = extended_F
+    if percolate_z:
+        extended_F = np.empty((F.shape[0], F.shape[1], 2 * F.shape[2]),
+                              dtype=F.dtype)
+        extended_F[:, :, :Z_real] = F
+        extended_F[:, :, Z_real:] = F
+        F = extended_F
+    X, Y, Z = F.shape
+
+    # Get F on a graph
+    F_graph = free_energy_graph(F, max_energy_threshold=1e7, diagonal=True)
+
+    # Find the lowest cost path that percolates along the x dimension
+    total_energy_cost = float('inf')
+    best_starting_point = None
+    best_perc_path = []
+
+    for starting_point in peaks:
+
+        # Get the end point which is a periodic image of the peak
+        end_point = (starting_point[0] + X_real * percolate_x,
+                     starting_point[1] + Y_real * percolate_y,
+                     starting_point[2] + Z_real * percolate_z)
```

Comment:
Define image above loop.

```suggestion
        end_point = starting_point + image
```

Reply:
done